pkgs/tools/misc/nbqa/default.nix

```diff
@@ -0,0 +1,90 @@
+{ black
+, blacken-docs
+, fetchFromGitHub
+, lib
+, python3
+, ruff
+}:
+python3.pkgs.buildPythonApplication rec {
+  pname = "nbqa";
+  version = "1.7.0";
+
+  src = fetchFromGitHub {
+    owner = "nbQA-dev";
+    repo = "nbQA";
+    rev = version;
+    hash = "sha256-CTF5HisBS44Ta18cVT04UrMIF30WmEBwZBGW7fkKXwk=";
+  };
+
+  passthru.optional-dependencies = {
+    black = [ black ];
+    blacken-docs = [ blacken-docs ];
+    flake8 = [ python3.pkgs.flake8 ];
+    isort = [ python3.pkgs.isort ];
+    jupytext = [ python3.pkgs.jupytext ];
+    mypy = [ python3.pkgs.mypy ];
+    pylint = [ python3.pkgs.pylint ];
+    pyupgrade = [ python3.pkgs.pyupgrade ];
+    ruff = [ ruff ];
+  };
+
+  propagatedBuildInputs = with python3.pkgs;
+    [
+      autopep8
+      ipython
+      tokenize-rt
+      tomli
+    ]
+    ++ builtins.attrValues passthru.optional-dependencies;
+
+  postPatch = ''
+    # Force using the Ruff executable rather than the Python package
+    substituteInPlace nbqa/__main__.py --replace 'if shell:' 'if shell or main_command == "ruff":'
+  '';
+
+  preCheck = ''
+    # Allow the tests to run `nbqa` itself from the path
+    export PATH="$out/bin":"$PATH"
+  '';
+
+  nativeCheckInputs = [
```

Comment:
> OK, I've removed with python3.pkgs; since that would make black ambiguous to the reader and would install the wrong package.

- `with` doesn't shadow keys that are already in scope. so top level black would still be used
- pkgs.black can be used here as well. top level black is same as pkgs.black except it cant be imported, which is not an issue here

you can split nativeCheckInputs into two concatenated lists, one `with python.pkgs;`

Reply:
I don't understand why it's OK to add so much complexity to save a few characters, but I've done as requested.

Reply:
appreciate your patience !

as for `nativeCheckInputs` splitting into two list was optional suggestion, I will try to me more explicit about suggestions in the future. 

derivation would still use correct (application) `black` if whole list had `with python.pkgs`. and actually would also work with `black` from `python.pkgs` (library)

as for duplication/deduplication using `with;`  eg  `[ a.b a.c ];` vs  `with a; [ b c ];` 
maybe we should add something to https://nixos.org/manual/nixpkgs/stable/#chap-conventions ? to make it easier for both authors and reviewers to apply conventions. maybe @mweinelt can suggest how we can improve this area.

otherwise, thank you taking your time, hopefully future PRs will become more straightforward

Reply:
nixpkgs has many maintainers, and a large number of PRs are submitted every day.
In order to maintain and review them continuously, it would be easier for both sides to follow a certain convention.
As long as it is not a bad practice, the opinion of the package maintainer can be respected, but there needs to be a good reason for doing something out of the convention.

Of course, undocumented practices are not good, and we need to try to resolve them.